crates/polars-lazy/src/physical_plan/expressions/ternary.rs

```diff
@@ -146,166 +121,168 @@ impl PhysicalExpr for TernaryExpr {
             (op_mask(), (op_truthy(), op_falsy()))
         };
 
-        let ac_mask = ac_mask?;
+        let mut ac_mask = ac_mask?;
         let mut ac_truthy = ac_truthy?;
         let mut ac_falsy = ac_falsy?;
 
-        let mask_s = ac_mask.flat_naive();
-
-        // BIG TODO: find which branches are never hit and remove them.
         use AggState::*;
-        match (ac_truthy.agg_state(), ac_falsy.agg_state()) {
-            // All branches are aggregated-flat or literal
-            // mask -> aggregated-flat
-            // truthy -> aggregated-flat | literal
-            // falsy -> aggregated-flat | literal
-            // simply align lengths and zip
-            (
-                Literal(truthy) | AggregatedScalar(truthy),
-                AggregatedScalar(falsy) | Literal(falsy),
-            )
-            | (AggregatedList(truthy), AggregatedList(falsy))
-                if matches!(ac_mask.agg_state(), AggState::AggregatedScalar(_)) =>
-            {
-                let mut truthy = truthy.clone();
-                let mut falsy = falsy.clone();
-                let mut mask = ac_mask.series().bool()?.clone();
-                expand_lengths(&mut truthy, &mut falsy, &mut mask);
-                let out = truthy.zip_with(&mask, &falsy).unwrap();
-                ac_truthy.with_series(out.with_name(truthy.name()), true, Some(&self.expr))?;
-                Ok(ac_truthy)
-            },
 
-            // We cannot flatten a list because that changes the order, so we apply over groups.
-            (AggregatedList(_), NotAggregated(_)) | (NotAggregated(_), AggregatedList(_)) => {
-                finish_as_iters(ac_truthy, ac_falsy, ac_mask)
-            },
+        let mut has_non_unit_literal = false;
+        for ac in [&mut ac_mask, &mut ac_truthy, &mut ac_falsy].into_iter() {
+            match ac.agg_state() {
+                Literal(s) => {
+                    has_non_unit_literal = s.len() != 1;
 
-            // Then:
-            //     col().shift()
-            // Otherwise:
-            //     None
-            (AggregatedList(_), Literal(_)) | (Literal(_), AggregatedList(_)) => {
-                if !aggregation_predicate {
-                    return finish_as_iters(ac_truthy, ac_falsy, ac_mask);
-                }
-                let mask = mask_s.bool()?;
-                let check_length = |ca: &ListChunked, mask: &BooleanChunked| {
-                    polars_ensure!(
-                        ca.len() == mask.len(), expr = self.expr, ComputeError:
-                        "predicates length: {} does not match groups length: {}",
-                        mask.len(), ca.len()
-                    );
-                    Ok(())
-                };
+                    if has_non_unit_literal {
+                        break;
+                    }
+                },
+                NotAggregated(_) => {
+                    let _ = ac.aggregated();
```

Comment:
I think before we start aggregating all columns we should handle the super cheap case, which is all non-aggregated or all unit literal and non-aggregated.